crates/sui/src/client_commands.rs

```diff
@@ -732,6 +732,17 @@ impl SuiClientCommands {
                 serialize_signed_transaction,
                 lint,
             } => {
+                if build_config.test_mode {
+                    return Err(SuiError::ModulePublishFailure {
+                        error: "The `publish` subcommand should not be used with the `--test` flag\n\
+                                \n\
+                                Library code in published packages must not depend on test code.\n\
+                                In order to fix this and publish the package without `--test`, search for, and remove \
+                                instances of e.g. test modules declared as `friend`s of modules from `sources/`.".to_string(),
```

Comment:
I would most likely rephrase the last part of this a bit as the way tests can be defined (and how modules could depend on other test-only code) can differ, e.g., you can have test-only modules under `sources/`, and even though this is an example at the end I worry it may put blinders on folks and how they look at fixing this issue.

What do you think of maybe something like:
```
In order to fix this and publish the package without `--test` remove any non-test dependencies on test-only code. You can ensure all test-only dependencies have been removed by compiling the package normally with `sui move build`. 
```

Reply:
Sounds good to me.